File: depthFirstSearch.js
class Node {
  constructor(name) {
    this.name = name;
    this.children = [];
  }

  addChild(name) {
    this.children.push(new Node(name));
    return this;
  }

	depthFirstHelper({name, children}, arr){
		arr.push(name)
		if(children.length) {
			children.forEach(child => {
				this.depthFirstHelper(child, arr)
			})
		}
	}
  
	//when function hits node add node name to result array
	//then see if a node has children if so call depthFirst on each child
  //apply helper to each node while passing the result array into each function call
  //array is finally return after all recursive calls are made
  
  depthFirstSearch(array) {
		this.depthFirstHelper(this, array)
		return array
  }
}

File: easy-problems/Python_solutions/depth_first_search.py
class Node:
    def __init__(self, name):
        self.children = []
        self.name = name

    def addChild(self, name):
        self.children.append(Node(name))
        return self

    def depthFirstSearch(self, array):
        depth_first_helper(self, array)
        return array


def depth_first_helper(node, arr):
    name = node.name
    children = node.children
    arr.append(name)
    if(len(children)):
        for i in children:
            depth_first_helper(i, arr)
